calico_node/startup.go

```diff
@@ -14,59 +14,657 @@
+	log.Info("Loading client config")
 	cfg, err := client.LoadClientConfig("")
 	if err != nil {
-		panic(fmt.Sprintf("Error loading config: %s", err))
+		fatal("Error loading datastore config: %s", err)
+		terminate()
 	}
-	c, err := backend.NewClient(*cfg)
+
+	log.Info("Creating client")
+	client, err := client.New(*cfg)
 	if err != nil {
-		panic(fmt.Sprintf("Error creating client: %s", err))
+		fatal("Error accessing the Calico datastore: %s", err)
+		terminate()
+	}
+
+	// An explicit value of false is required to skip waiting for the
+	// datastore.
+	if os.Getenv("WAIT_FOR_DATASTORE") == "false" {
+		message("Skipping datastore connection test")
+		return cfg, client
+	}
+
+	message("Checking datastore connection")
+	for i := 0; i <= 120; i++ {
+		// Query some arbitrary configuration to see if the connection
+		// is working.  Getting a specific node is a good option, even
+		// if the node does not exist.
+		_, err = client.Nodes().Get(api.NodeMetadata{Name: "foo"})
+
+		// We only care about a couple of error cases, all others would
+		// suggest the datastore is accessible.
+		if err != nil {
+			switch err.(type) {
+			case errors.ErrorConnectionUnauthorized:
+				fatal("Connection to the datastore is unauthorized")
+				terminate()
+			case errors.ErrorDatastoreError:
+				time.Sleep(1000 * time.Millisecond)
+				continue
+			}
+		}
+
+		err = nil
+		break
 	}
-	log.Info("Ensuring datastore is initialized")
-	err = c.EnsureInitialized()
 	if err != nil {
-		panic(fmt.Sprintf("Error initializing datastore: %s", err))
+		fatal("Connection to the datastore has failed: %s", err)
+		terminate()
 	}
-	log.Info("Datastore is initialized")
 
-	// Make sure we have a global cluster ID set.
-	log.Info("Ensuring a cluster guid is set")
-	ensureClusterGuid(c)
+	message("Datastore connection verified")
+
+	return cfg, client
 }
 
-// ensureClusterGuid assigns a cluster GUID if one doesn't exist.
-func ensureClusterGuid(c api.Client) {
-	guid := hex.EncodeToString(uuid.NewV4().Bytes())
-	_, err := c.Create(&model.KVPair{
-		Key:   model.GlobalConfigKey{Name: "ClusterGUID"},
-		Value: guid,
-	})
+// writeStartupEnv writes out the startup.env file to set environment variables
+// that are required by confd/bird etc. but may not have been passed into the
+// container.
+func writeStartupEnv(nodeName string, ip string, ip6 string) {
+	text := "export HOSTNAME=" + nodeName + "\n"
+	text += "export NODENAME=" + nodeName + "\n"
+	text += "export IP=" + ip + "\n"
+	text += "export IP6=" + ip6 + "\n"
+
+	// Write out the startup.env file to ensure require environments are
+	// set (which they might not otherwise be).
+	if err := ioutil.WriteFile("startup.env", []byte(text), 0666); err != nil {
+		log.WithError(err).Info("Unable to write to startup.env")
+		fatal("Unable to write to local filesystem")
+		terminate()
+	}
+}
+
+// getNode returns the current node configuration and a list of all configured
```

Comment:
well I may have changed this at some point and not updated the comment..... whistle .. la la la